lib/resty/rediscluster.lua

```diff
@@ -76,6 +76,25 @@ local function check_auth(self, redis_client)
             return nil, err
         end
 
+    -- redis 6.x adds support for username+password combination
+    elseif type(self.config.password) == "string" then
+        local count, err = redis_client:get_reused_times()
+        if count == 0 then
+            local _
+
+            if type(self.config.username) == "string" then
+                _, err = redis_client:auth(self.config.username, self.config.password)
+            else
+                _, err = redis_client:auth(self.config.password)
+            end
+        end
+
+        if not err then
+            return true, nil
+        else
+            return nil, err
+        end
```

Comment:
I think this logic would be easier to follow if the `err` would not be reassigned, but instead the function would return at the earliest point in time.  This seems to be a security relevant function that needs to be easy to review, so a refactoring is in order.

Reply:
Not sure what I should change here - it's only a little messy because I wanted to retain the backwards compatibility, if the config table is passed in with only "auth" field set.

What should I do?

Reply:
@hanshuebner though I agree with you, the logic is identical to the check just before it (existing code).

Reply:
```
local count, err = redis_client:get_reused_times()
if err then
  return nil, err
end

if count ~= 0 then
  # why?
  return true
end

if type(self.config.username) == "string" then
  return redis_client:auth(self.config.username, self.config.password)
end

# if we don't have a username, we just use the password (?)
return redis_client:auth(self.config.password)
```
This, and a couple of comments, would go a long way.  While at it, clarify the other code path as well.

Reply:
In this place, we should check if the `username` is *default* or not. If it is set to *default*, we should return error.

Reply:
my take;
```lua
local count, err = redis_client:get_reused_times()
if not count then
  return nil, err
elseif count ~= 0 then
  return true -- reusing connection, no need to authenticate again
end

if type(self.config.username) == "string" then
  return redis_client:auth(self.config.username, self.config.password)
end

# if we don't have a username, we just use the password (?)
return redis_client:auth(self.config.password)
```

Reply:
`default` is what redis 6.0+ now uses to refer to the old style "password only" user - so `default` is automatically assumed when ONLY a password is sent, with no accompanying username. It's a little weird.

Usually I would say just accept if the administrator puts "default" into the username box? Why not?!

Reply:
> `default` is what redis 6.0+ now uses to refer to the old style "password only" user - so `default` is automatically assumed when ONLY a password is sent, with no accompanying username. It's a little weird.
> 
> Usually I would say just accept if the administrator puts "default" into the username box? Why not?!

As long as a user provides a `username` and `password` combination, we should do sanity check against the value of `username`.

Reply:
Is it intentional that the code falls through if `count ~= null` or if username and password authentication return no err?  Generally, things are easier to understand when `if` statements are not deeply nested.  If a condition exists that allows returning, return.

Reply:
> @outsinre any docs on that 'default' value? (I'd like to understand)

Here is a reference link: https://docs.konghq.com/hub/kong-inc/rate-limiting-advanced/

Check the `config.redis.username` part.